xapian-core/api/omenquire.cc

```diff
@@ -452,6 +464,28 @@ MSet::Internal::read_docs() const
     requested_docs.clear();
 }
 
+void
+MSet::Internal::set_item_weight(int i, double wt_)
+{
+    set_item_weight_(items[i], wt_);
+    if(i == 0)
+	max_attained = wt_;
+    max_attained = max(max_attained, wt_);
+}
+
+void
+MSet::Internal::set_item_weight_(Xapian::Internal::MSetItem& item_, double wt_)
+{
+    item_.wt = wt_;
+}
+
+void
+MSet::Internal::sort_by_relevance()
+{
+    std::sort(items.begin(),items.end(),
+	      [](const Xapian::Internal::MSetItem& x, const Xapian::Internal::MSetItem& y){ return x.wt > y.wt; });
```

Comment:
Yes, though actually this should fall back to comparing by docid if the weights are equal so that this always provides a stable sort order - e.g. look at the comparisons in `matcher/msetcmp.cc`:

```
// Order by relevance, then docid.
template<bool FORWARD_DID> bool
msetcmp_by_relevance(const Xapian::Internal::MSetItem &a,
                     const Xapian::Internal::MSetItem &b)
{
    if (a.wt > b.wt) return true;
    if (a.wt < b.wt) return false;
    return msetcmp_by_did<FORWARD_DID, true>(a, b);
}
```

It probably actually makes sense to just use `get_mset_cmp_function(Xapian::Enquire::Internal::REL, true, false)` to get a suitable comparator which does that and which we know works consistently with standard `MSet` sorting (the `true` hard wires the forward/reverse ordering of docids - I don't think it matters to respect the order specified when the `MSet` was created here, though we could get it from the `Enquire` object which the `MSet` has a reference to).